sns_ik_lib/matlab/snsIk_acc_rr_cs.m

```diff
@@ -0,0 +1,203 @@
+%   sCS = nullspace configuration space (CS) task scale
+%   exitCode    (1 == success)
+%
+%
+% NOTES:
+%   s * ddxGoal = J * ddq + dJ * dq;
+%   sCS * ddqCS = (I - pinv(J)*J) * ddq;
+%
+% This function is a modification of the basic SNS-IK algorithm that is
+% presented in the original SNS-IK papers. It was developed by Andy Park at
+% Rethink Robotics in June 2018.
+%
+
+% Copyright 2018 Rethink Robotics
+%
+% Licensed under the Apache License, Version 2.0 (the "License");
+% you may not use this file except in compliance with the License.
+% You may obtain a copy of the License at
+% http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS,
+% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+% See the License for the specific language governing permissions and
+% limitations under the License.
+%
+
+% TODO: input validation
+% TODO: return optimization status
+
+%% compute the solution for the primary task
+[nTask, nJnt] = size(J);
+W = eye(nJnt);
+ddqNull = zeros(nJnt, 1);
+s = 1.0;
+sStar = 0.0;
+exitCode = 1;
+tol = 1e-6;
+limitExceeded = true;
+while limitExceeded == true
+    limitExceeded = false;
+    ddq = ddqNull + pinv(J*W) * (ddxGoal -dJdq - J*ddqNull);
+
+    if any(ddq < (ddqLow - tol)) || any(ddq > (ddqUpp + tol))
+        limitExceeded = true;
+    end
+    a = pinv(J*W) * ddxGoal;
+    b = ddq - a;
+
+    marginL = ddqLow - b;
+    marginU = ddqUpp - b;
+    sMax = zeros(nJnt, 1);
+
+    for i=1:nJnt
+        if W(i,i) == 0
+            sMax(i) = inf;
+        else
+            sMax(i) = FindScaleFactor(marginL(i), marginU(i), a(i));
+            % if scale factor is 1 but joint limit is violated,
+            % set a scale factor to a small number so that the
+            % corresponding joint will be saturated
+            if(sMax(i) == 1 && (ddq(i) < (ddqLow(i) - tol) || ddq(i) > (ddqUpp(i) + tol)))
+                sMax(i) = 1e-3;
+            end
+        end
+    end
+
+    [~, jntIdx] = min(sMax);
+    taskScale = sMax(jntIdx);
+    % if the current best scaled solution violates the limit,
+    % update sStar to the latest scale factor
+    ddqTmp = ddqNull + pinv(J*W)*(sStar * ddxGoal - dJdq - J * ddqNull);
+    if ((taskScale > sStar) || (any(ddqTmp > ddqUpp + tol) || any(ddqTmp < ddqLow - tol)))
+        sStar = taskScale;
+        Wstar = W;
+        ddqNullStar = ddqNull;
+    end
+
+    W(jntIdx, jntIdx) = 0;
+    ddqNull(jntIdx) = min(max(ddqLow(jntIdx), ddq(jntIdx)), ddqUpp(jntIdx));
+
+    if rank(J*W) < nTask
+        s = sStar;
+        W = Wstar;
+        ddqNull = ddqNullStar;
+        ddq = ddqNull + pinv(J*W)*(s * ddxGoal - dJdq - J * ddqNull);
+        limitExceeded = false;
+    end
+end
+
+%% compute the solution for the secondary task
+%-- initialization
+ddq1 = ddq;
+I = eye(nJnt);
+Wcs = I;
+P1 = (I - pinv(J)*J);
+
+%-- start of the algorithm
```

Comment:
Add more detail here. I think this is the algorithm for handling the secondary task. Also put a note about how this might be extended to an arbitrary number of tasks.

Reply:
For example, I think this is very close to unwrapping a for loop with two iterations.

Reply:
I added more description. I am going to commit a generic algorithm as well today.